rules/ruleset.go

```diff
@@ -0,0 +1,355 @@
+		}
+		if cutoverNs < snapshot.cutoverNs {
+			cutoverNs = snapshot.cutoverNs
+		}
+		for _, target := range snapshot.targets {
+			found := false
+			// If the new target has the same transformation as an existing one,
+			// we merge their policies.
+			for i := range rollups {
+				if rollups[i].sameTransform(target) {
+					rollups[i].Policies = append(rollups[i].Policies, target.Policies...)
+					found = true
+					break
+				}
+			}
+			// Otherwise, we add a new rollup target.
+			if !found {
+				rollups = append(rollups, target.clone())
+			}
+		}
+	}
+
+	// Resolve the policies for each rollup target.
+	for i := range rollups {
+		rollups[i].Policies = resolvePolicies(rollups[i].Policies)
+	}
+
+	return cutoverNs, as.toRollupResults(id, rollups)
+}
+
+// toRollupResults encodes rollup target name and values into ids for each rollup target.
+func (as *activeRuleSet) toRollupResults(id []byte, targets []rollupTarget) []rollupResult {
+	// NB(r): This is n^2 however this should be quite fast still as
+	// long as there is not an absurdly high number of rollup
+	// targets for any given ID and that iterfn is alloc free.
+	//
+	// Even with a very high number of rules its still predicted that
+	// any given ID would match a relatively low number of rollups.
+
+	// TODO(xichen): pool tag pairs and rollup results.
+	var tagPairs []TagPair
+	rollups := make([]rollupResult, 0, len(targets))
+	for _, target := range targets {
+		tagPairs = tagPairs[:0]
+		for _, tag := range target.Tags {
+			iter := as.iterFn(id)
+			for iter.Next() {
+				name, value := iter.Current()
+				if bytes.Equal(name, tag) {
+					tagPairs = append(tagPairs, TagPair{Name: tag, Value: value})
+					break
+				}
+			}
+			iter.Close()
+		}
+		result := rollupResult{
+			ID:       as.newIDFn(target.Name, tagPairs),
+			Policies: target.Policies,
+		}
+		rollups = append(rollups, result)
+	}
+
+	return rollups
+}
+
+// nextCutover returns the next cutover time after t.
+// NB(xichen): not using sort.Search to avoid a lambda capture.
+func (as *activeRuleSet) nextCutover(t int64) int64 {
+	i, j := 0, len(as.cutoverTimesAsc)
+	for i < j {
+		h := i + (j-i)/2
+		if as.cutoverTimesAsc[h] <= t {
+			i = h + 1
+		} else {
+			j = h
+		}
+	}
+	if i < len(as.cutoverTimesAsc) {
+		return as.cutoverTimesAsc[i]
+	}
+	return timeMax
+}
+
+// RuleSet is a set of rules associated with a namespace.
+type RuleSet interface {
+	// Namespace is the metrics namespace the ruleset applies to.
+	Namespace() string
+
+	// Version returns the ruleset version.
+	Version() int
+
+	// CutoverNs returns when the ruleset takes effect.
+	CutoverNs() int64
+
+	// TombStoned returns whether the ruleset is tombstoned.
+	TombStoned() bool
+
+	// ActiveSet returns the active ruleset at a given time.
+	ActiveSet(t time.Time) Matcher
+}
+
+type ruleSet struct {
+	uuid            string
+	version         int
+	namespace       string
+	createdAtNs     int64
+	lastUpdatedAtNs int64
+	tombStoned      bool
```

Comment:
nit: This should just be `tombstoned` to match the proto def and also the dictionary definition. Tombstone is a single word.